File: src/golang/03/solution_03.go
// The prime factors of 13195 are 5, 7, 13 and 29.
// What is the largest prime factor of the number 600851475143?
// Answer: 6857

package main

import (
	"flag"
	"fmt"
	"math"
)

var (
	number = flag.Int64("number", 600851475143, "input number")
)

func init() {
	flag.Parse()
}

func isPrime(num int64) bool {
	if num <= 1 {
		return false
	}
	if num == 2 || num == 3 {
		return true
	}

	sqrtValue := int64(math.Sqrt(float64(num)))
	for i := int64(4); i <= sqrtValue; i++ {
		if num%i == 0 {
			return false
		}
	}

	return true
}

func method0(number int64) (res int64) {
	sqrtValue := int64(math.Sqrt(float64(number)))
	for i := sqrtValue; i >= 2; i-- {
		if number%i == 0 && isPrime(i) {
			return i
		}
	}

	return -1
}

func method1(number int64) (res int64) {
	for i := int64(2); i <= number/2; i++ {
		if number%i == 0 {
			number = number / i
			if isPrime(number) {
				return number
			}
			method1(number)
		}
	}

	return -1
}

func main() {
	fmt.Println(method0(*number))
	fmt.Println(method1(*number))
}

File: src/python/solution_03.py
"""
The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143 ?

Answer: 6857
"""

import math


class Solution0:

    def is_prime(self, num):
        if not isinstance(num, int) or num <= 1:
            return False

        end = int(math.sqrt(num))
        for i in range(2, end+1):
            if num % i == 0:
                return False

        return True

    def get_max_prime(self, num=600851475143):
        end = int(math.sqrt(num))
        for i in range(end, 1, -1):
            if num % i == 0 and self.is_prime(i):
                return i

        return -1


class Solution1:

    def is_prime(self, num):
        if not isinstance(num, int) or num <= 1:
            return False

        end = int(math.sqrt(num))
        for i in range(2, end+1):
            if num % i == 0:
                return False

        return True

    def get_max_prime(self, num=600851475143):
        for i in range(2, int(num/2)):
            if num % i == 0:
                num = int(num/i)
                if self.is_prime(num):
                    return num

                self.get_max_prime(num)


if __name__ == '__main__':
    solution_0 = Solution0()
    print(solution_0.get_max_prime())

    solution_1 = Solution1()
    print(solution_1.get_max_prime())
